File: frontend/types/scoring.ts
import type { BusinessScenario } from "@/types/scenario";

export type OpportunityLevel = "low" | "medium" | "high" | "priority";

export type UnitScoringResult = {
  id: string;
  plant_id: string;
  scenario_id: string;
  scenario_result_id: string;
  scoring_run_id: string;
  opportunity_score: number;
  readiness_score: number;
  confidence_score: number;
  composite_score: number;
  co2_availability_score: number;
  methanol_potential_score: number;
  h2_readiness_score: number;
  economic_return_score: number;
  infrastructure_score: number;
  land_port_score: number;
  market_access_score: number;
  risk_permit_score: number;
  data_completeness_score: number;
  emission_data_quality_score: number;
  land_readiness_score: number;
  utility_readiness_score: number;
  h2_strategy_clarity_score: number;
  permit_logistic_readiness_score: number;
  carbon_credit_potential_score: number;
  strategic_value_score: number;
  utility_advantage_score: number;
  component_scores: Record<string, unknown>;
  data_gap_count: number;
  rank_position: number | null;
  recommended_scheme: string;
  key_bottleneck: string;
  scoring_version: string;
  created_at: string;
};

export type ScoringRecalculateResponse = {
  scoring_run_id: string | null;
  created_count: number;
  results: UnitScoringResult[];
};

export type UnitRankingRow = {
  rank: number;
  plant_id: string;
  site_id: string;
  scenario_id: string;
  scenario_result_id: string;
  scoring_result_id: string;
  scenario_name: string;
  scheme: string;
  site_name: string;
  unit_name: string;
  province: string | null;
  city: string | null;
  latitude: number | null;
  longitude: number | null;
  capacity_mw: number | null;
  fuel_type: string | null;
  composite_score: number;
  opportunity_score: number;
  readiness_score: number;
  confidence_score: number;
  heatmap_weight: number;
  co2_tpy: number | null;
  captured_co2_tpy: number | null;
  methanol_tpy: number | null;
  h2_required_tpy: number | null;
  estimated_irr: number | null;
  estimated_lcom_usd_ton: number | null;
  recommended_scheme: string;
  key_bottleneck: string;
  data_gap_count: number;
  data_confidence_label: string;
  opportunity_level: OpportunityLevel;
  calculation_version: string;
};

export type UnitOpportunityFeature = {
  type: "Feature";
  id: string;
  geometry: {
    type: "Point";
    coordinates: [number, number];
  };
  properties: {
    site_id: string;
    site_name: string;
    unit_name: string;
    province: string | null;
    city: string | null;
    capacity_mw: number | null;
    fuel_type: string | null;
    scenario_id: string;
    scenario_name: string;
    scheme: string;
    opportunity_score: number;
    readiness_score: number;
    confidence_score: number;
    composite_score: number;
    co2_tpy: number | null;
    captured_co2_tpy: number | null;
    methanol_tpy: number | null;
    h2_required_tpy: number | null;
    estimated_irr: number | null;
    estimated_lcom_usd_ton: number | null;
    rank_position: number;
    data_confidence_label: string;
    heatmap_weight: number;
    recommended_scheme: string;
    key_bottleneck: string;
    data_gap_count: number;
    opportunity_level: OpportunityLevel;
    status: string;
  };
};

export type UnitOpportunityGeoJSON = {
  type: "FeatureCollection";
  features: UnitOpportunityFeature[];
};

export type UnitProfile = {
  plant: Record<string, unknown>;
  scenario: Partial<BusinessScenario> | null;
  latest_result: Record<string, unknown> | null;
  latest_scoring: Record<string, unknown> | null;
  score_breakdown: Record<string, unknown>;
  data_gaps: Array<{
    module: string;
    name: string;
    impact: string;
    priority: string;
  }>;
  confidence_labels: Record<string, string>;
};

export type DashboardFilters = {
  scenarioId: string;
  scheme: string;
  region: string;
  fuelType: string;
  confidence: string;
  opportunityLevel: string;
  showHeatmap: boolean;
  showMarkers: boolean;
  showLabels: boolean;
};

File: backend/app/schemas/scoring.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


OpportunityLevel = Literal["low", "medium", "high", "priority"]


class ScoringRecalculateRequest(BaseModel):
    scenario_id: str | None = None
    scheme: str | None = None


class UnitScoringRead(BaseModel):
    id: str
    plant_id: str
    scenario_id: str
    scenario_result_id: str
    scoring_run_id: str
    opportunity_score: float
    readiness_score: float
    confidence_score: float
    composite_score: float
    co2_availability_score: float
    methanol_potential_score: float
    h2_readiness_score: float
    economic_return_score: float
    infrastructure_score: float
    land_port_score: float
    market_access_score: float
    risk_permit_score: float
    data_completeness_score: float
    emission_data_quality_score: float
    land_readiness_score: float
    utility_readiness_score: float
    h2_strategy_clarity_score: float
    permit_logistic_readiness_score: float
    carbon_credit_potential_score: float
    strategic_value_score: float
    utility_advantage_score: float
    component_scores: dict[str, Any] = Field(default_factory=dict)
    data_gap_count: int
    rank_position: int | None
    recommended_scheme: str
    key_bottleneck: str
    scoring_version: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ScoringRecalculateResponse(BaseModel):
    scoring_run_id: str | None
    created_count: int
    results: list[UnitScoringRead]


class UnitRankingRow(BaseModel):
    rank: int
    plant_id: str
    site_id: str
    scenario_id: str
    scenario_result_id: str
    scoring_result_id: str
    scenario_name: str
    scheme: str
    site_name: str
    unit_name: str
    province: str | None
    city: str | None
    latitude: float | None
    longitude: float | None
    capacity_mw: float | None
    fuel_type: str | None
    composite_score: float
    opportunity_score: float
    readiness_score: float
    confidence_score: float
    heatmap_weight: float
    co2_tpy: float | None
    captured_co2_tpy: float | None
    methanol_tpy: float | None
    h2_required_tpy: float | None
    electrolyzer_required_mw: float | None
    gross_revenue_usd_per_year: float | None
    estimated_irr: float | None
    estimated_lcom_usd_ton: float | None
    recommended_scheme: str
    key_bottleneck: str
    data_gap_count: int
    data_confidence_label: str
    opportunity_level: OpportunityLevel
    calculation_version: str


class GeoJSONGeometry(BaseModel):
    type: Literal["Point", "LineString", "Polygon", "MultiPoint", "MultiLineString", "MultiPolygon"]
    coordinates: Any


class GeoJSONFeature(BaseModel):
    type: Literal["Feature"]
    id: str
    geometry: GeoJSONGeometry
    properties: dict[str, Any]


class GeoJSONFeatureCollection(BaseModel):
    type: Literal["FeatureCollection"]
    features: list[GeoJSONFeature]


class UnitProfileRead(BaseModel):
    plant: dict[str, Any]
    scenario: dict[str, Any] | None
    latest_result: dict[str, Any] | None
    latest_scoring: dict[str, Any] | None
    score_breakdown: dict[str, Any]
    data_gaps: list[dict[str, str]]
    confidence_labels: dict[str, str]
